winsup/utils/getprocaddr.c

```diff
@@ -0,0 +1,198 @@
+#include <stdio.h>
+#include <windows.h>
+
+/**
+ * To determine the address of kernel32!CtrlRoutine, we need to use
+ * dbghelp.dll. But we want to avoid linking statically to that library because
+ * the normal operation of cygwin-console-helper.exe (i.e. to allocate a hidden
+ * Console) does not need it.
+ *
+ * Therefore, we declare the SYMBOL_INFOW structure here, load the dbghelp
+ * library via LoadLibraryExA() and obtain the SymInitialize(), SymFromAddrW()
+ * and SymCleanup() functions via GetProcAddr().
+ * 
+ * After that a remote thread is injected into that process
+ * to send the exit code.
+ */
+
+#include <dbghelp.h>
+
+static int pid;
+static uintptr_t exit_code;
+static BOOL do_print_address;
+
+/* Avoid fprintf(), as it would try to reference '__getreent' */
```

Comment:
Is this still true?

Reply:
No idea...

Reply:
Could I ask you to replace that `WriteFile()` dance by an `fprintf()` and see whether the program links fine? If it does, then you don't need that super ugly `output()` function anymore.

Reply:
I did something like
```patch
diff --git a/winsup/utils/getprocaddr.c b/winsup/utils/getprocaddr.c
index 0fa07827e..133c0bbfb 100644
--- a/winsup/utils/getprocaddr.c
+++ b/winsup/utils/getprocaddr.c
@@ -32,7 +32,7 @@ output (BOOL error, const char *fmt, ...)
   vsnprintf (buffer, sizeof(buffer) - 1, fmt, ap);
   buffer[sizeof(buffer) - 1] = '\0';
   va_end (ap);
-  WriteFile (GetStdHandle(error ? STD_ERROR_HANDLE : STD_OUTPUT_HANDLE),
+  fprintf (error ? stderr : stdout,
              buffer, strlen (buffer), NULL, NULL);
 }
```
and it seems to work.

Reply:
ah, this does output the pointer. no idea why.

Reply:
Could you help me understand what you mean by "the pointer"?

Reply:
I mean, when doing 
```sh
$ ./getprocaddr CtrlRoutine
0000000077b54810
Segmentation fault
```
the character prints, I meant that as the pointer. When I try the same with that patch it doesn't print anything.

Reply:
Oh.

In any case, I think we can totally side-step this problem by using `printf()` and `fprintf()` directly.

Reply:
I think we can use `printf` and `fprintf` wherever possible and let the `output` function only for the pointer.

Reply:
We do not need `output()` for the pointer, we can easily call `printf("%p", address)`.

Reply:
I wonder whether it will get printed correctly after we have shifted to a new console(using `AllocConsole`), but maybe I will try that and then see.

Reply:
As I expected, the pointer doesn't print in the original console, rather it prints on the popup and then exits. Atleast, when using `output()` it printed on the original console on the MSYS shell (on cmd it never printed in the original shell.).

Reply:
I did https://github.com/msys2/msys2-runtime/pull/31/commits/b1c94c16f33267df692b776c4797a2e93ff96b32, with that the pointer for `CtrlRoutine` doesn't print while `ExitProcess` prints.

Reply:
Good point! I hope the `GetStdHandle()`/`SetStdHandle()` dance I described elsewhere will work around that.